File: sdk/typescript/src/lib/handlers/FolderHandler.ts
import { MessageHandler } from './MessageHandler';
import { WebSocket } from 'ws';
import { WebSocketMessage } from '../WebSocketRouter';
import { LibraryServerDataSQLite } from '../LibraryServerDataSQLite';

export class FolderHandler extends MessageHandler {
  constructor(
    dbService: LibraryServerDataSQLite,
    ws: WebSocket,
    message: WebSocketMessage
  ) {
    super(dbService, ws, message);
  }

  async handle(): Promise<void> {
    try {
      const { action, payload } = this.message;
      const { data } = payload;
      
      let result;
      switch(action) {
        case 'query':
          result = await this.dbService.queryFolder(data.query);
          break;
        case 'create':
          result = await this.dbService.createFolder(data);
          break;
        case 'update':
          result = await this.dbService.updateFolder(data.id, data);
          break;
        case 'delete':
          result = await this.dbService.deleteFolder(data.id);
          break;
        default:
          throw new Error(`Unsupported folder action: ${action}`);
      }

      this.sendResponse({data: result});
    } catch (err) {
      this.sendError(err instanceof Error ? err.message : 'Folder operation failed');
    }
  }
}

File: lib/plugins/libraries/services/websocket_handlers/folder_handler.dart
import 'package:mira/plugins/libraries/services/websocket_handlers/common_handler.dart';
import 'package:mira/plugins/libraries/services/server_item_event.dart';

class FolderHandler extends BaseWebSocketHandler {
  final String action;
  final Map<String, dynamic> payload;

  FolderHandler({
    required super.dbService,
    required super.channel,
    required super.data,
    required super.requestId,
    required super.libraryId,
    required this.action,
    required this.payload,
  });

  @override
  Future<void> handle() async {
    try {
      switch (action) {
        case 'all':
          sendSuccess({'folders': await dbService.getAllFolders()});
          break;
        case 'read':
          if (payload.containsKey('id')) {
            final record = await dbService.getFolder(payload['id'] as int);
            sendSuccess({'record': record});
          } else {
            final records = await dbService.getFolders(
              limit: payload['limit'] as int? ?? 100,
              offset: payload['offset'] as int? ?? 0,
            );
            sendSuccess({'records': records});
          }
          break;

        case 'create':
          final id = await dbService.createFolder(data);
          final folders = await dbService.getAllFolders();
          sendSuccess({'id': id, 'folders': folders});
          break;

        case 'update':
          final success = await dbService.updateFolder(
            data['id'] as int,
            data['data'] as Map<String, dynamic>,
          );
          if (success) {
            final folders = await dbService.getAllFolders();
            sendSuccess({'folders': folders});
          } else {
            sendError('Update failed');
          }
          break;

        case 'delete':
          final success = await dbService.deleteFolder(data['id'] as int);
          if (success) {
            final folders = await dbService.getAllFolders();
            sendSuccess({'folders': folders});
          } else {
            sendError('Delete failed');
          }
          break;

        default:
          sendError('Unknown action for folders');
      }
    } catch (e) {
      sendError('Folder operation failed', e.toString());
    }
  }
}
